File: 78.subsets.java
/*
 * @lc app=leetcode id=78 lang=java
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> subsets(int[] nums) {
        if (nums.length == 0) return res;
        backtrack(0, nums, new ArrayList<>());
        return res;
    }

    private void backtrack(int s, int[] nums, List<Integer> tmpList) {
        res.add(new ArrayList<>(tmpList));
        for (int i = s; i < nums.length; ++i) {
            tmpList.add(nums[i]);
            backtrack(i+1, nums, tmpList);
            tmpList.remove(tmpList.size()-1);
        }
    }
}
// @lc code=end

File: go/78.subsets.go
/*
 * @lc app=leetcode id=78 lang=golang
 *
 * [78] Subsets
 */

// @lc code=start
func subsets(nums []int) [][]int {
	res := [][]int{}
	var backtrack func(int, []int)
	backtrack = func(idx int, comb []int) {
		tmp := make([]int, len(comb))
		copy(tmp, comb)
		res = append(res, tmp)
		
		for i := idx; i < len(nums); i++ {
			comb = append(comb, nums[i])
			backtrack(i+1, comb)
			comb = comb[:len(comb)-1]
		}
	}

	backtrack(0, []int{})
	return res
}
// @lc code=end

